### ISSUE TITLE
restart from vlsv does not work

### ISSUE THREAD

**Issue by rjarvinen** [fmihpc/corsair#2]
restarting rhybrid from a vlsv file produces errors:
(VLSV) ERROR: VLSV::getMPIDatatype called with datatype::UNKNOWN datatype, returning MPI_DATATYPE_NULL!

Not sure yet where the problem is: vlsv, corsair or rhybrid user code.

**Comment by sandroos** [comment]
It's https://github.com/fmihpc/vlsv/blob/master/vlsv_common_mpi.cpp#L104 that is failing, you could do a quick check of the VLSV file footer and check that all arrays have sensible datatype and byte size.

**Comment by rjarvinen** [comment]
Yes, that's it. particle arrays have "unknown" datatype in the restart file.
```
<DYNAMIC arraysize="29311" datasize="56" datatype="unknown" mesh="SpatialGrid" name="sw_H+_particles" type="celldata" vectorsize="1">1199269</DYNAMIC>
<DYNAMIC arraysize="1418" datasize="56" datatype="unknown" mesh="SpatialGrid" name="sw_He++_particles" type="celldata" vectorsize="1">2847597</DYNAMIC>
<DYNAMIC arraysize="3067" datasize="56" datatype="unknown" mesh="SpatialGrid" name="iono_O+_particles" type="celldata" vectorsize="1">2933917</DYNAMIC>
<DYNAMIC arraysize="3794" datasize="56" datatype="unknown" mesh="SpatialGrid" name="iono_O2+_particles" type="celldata" vectorsize="1">3112581</DYNAMIC>
<DYNAMIC arraysize="1099" datasize="56" datatype="unknown" mesh="SpatialGrid" name="exo_H+_particles" type="celldata" vectorsize="1">3331957</DYNAMIC>
<DYNAMIC arraysize="1841" datasize="56" datatype="unknown" mesh="SpatialGrid" name="exo_O+_particles" type="celldata" vectorsize="1">3400413</DYNAMIC>
```

**Comment by sandroos** [comment]
I don't have any code compiled at the moment nor a virtualbox installed, so I can't quickly check this myself. Do you have any old restart files around? I think those particle pops have always had 'unknown' datatype, so the change must be in vlsv.

**Comment by sandroos** [comment]
Hi, sorry for the delay, I found the issue in VLSV library. I'll have a pull request ready soon so that you can test it. I can't promise 100% that the old restart data is useful, however.

**Comment by sandroos** [comment]
Please try this https://github.com/fmihpc/vlsv/pull/32

**Comment by sandroos** [comment]
Marked this as 'help wanted' and 'wontfix' as the issue is in VLSV library and tracked by issue https://github.com/fmihpc/vlsv/issues/31

**Comment by rjarvinen** [comment]
Some update on the breakpointing issue. I am using https://github.com/fmihpc/vlsv/pull/32 patch for VLSV.

Starting a Corsair/RHybrid run from a breakpoint file results sometimes in few corrupted particles in particle lists. The corruptions include floating point quantities of a particle (x,y,z,vx,vy,vz,w) mixed to each other (e.g. vx gets the value of w or vice versa) or uninitialized floating points values (1e-300 etc).

These corrupted particles are not present in particle lists when restart files are written in the original run. I believe they must occur in the writing or reading the restart file. Not sure if it happens because of Corsair or VLSV.

I have tried applying updatePartitioning function of particle_list_skeleton.h just before writeRestart is called. This did not help. The corrupted particles associated with a restart seem to occur less frequently (or maybe not at all) if the mesh is not repartitioned during the original run.

I think one possible issue could arise if a SIZE(DYNAMIC) array would include for some reason a wrong number of particles in a cell/block. Or some other issue with writing or reading dynamic user arrays by Corsair. Or maybe this is still related to the bytesize issue, which was fixed in the VLSV "corsair_restart_fix" patch https://github.com/fmihpc/vlsv/pull/32.

I have not been able to construct a (very) minimal example outside of our Cray system yet. So this is not an easy one to debug.

![run03_crop](https://user-images.githubusercontent.com/7676607/41533279-e98da736-7302-11e8-9820-bf046abc5c24.png)

Figure caption: An example of how the corrupted particles are distributed by the MPI processes. On the right is the particle density after restarting with corrupted particles removed (by this method: https://github.com/fmihpc/rhybrid/commit/f85cb773243f1bc9d31aaf1de5a8a14e784a813b) and on the left are the MPI process ranks.

**Comment by smallrainbow** [comment]
Dear developers,

I meet the same problem as @rjarvinen. @sandroos, may I ask did this problem be solved?

Restart from the 1st restart file: restart00000000.vlsv is fine and the run can be well continued. However, if we restart from a later restart file: e.g., restart00000080.vlsv. The run reports as follows:

(VLSV) ERROR: VLSV::getMPIDatatype called with datatype::UNKNOWN datatype, returning MPI_DATATYPE_NULL!

Test runs have been carried out by openmpi1.8.8, gcc 8.4.1, and centOS8/Ubuntu20.04
Thank you very much!
